### PR TITLE
Initial LMDB BlockchainBackend implementation

### PR DESCRIPTION
**Pull request by neonknight64**
## Description
- Added LMDBDatabase chain storage backend with helper functions to perform lmdb operations
- Added LMDBVec that can be used as a persistent backend for the MerkleChangeTracker
- A set of tests for testing only the backend functionality of memory_db and lmdb_db was added under chain_backend tests
- A custom serializer  and deserializer was added for MerkeCheckPoint, this was required as croaring bitmap doesn't implement serialize and deserialize.

Note: The persistent storage of the MerkleChangeTrackers needs to be improved so that only a single WriteTransaction is used, this will also allow a single lmdb commit to be performed once all the changes have successfully been set.

## Motivation and Context
The Blockchain state requires persistent storage

## How Has This Been Tested?
A set of backend tests were added.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/chain_storage/memory_db.rs [hunk 1]
```diff
@@ -142,6 +142,7 @@ where D: Digest + Send + Sync
                     DbKey::Metadata(_) => {}, // no-op
                     DbKey::BlockHeader(k) => {
                         db.headers.remove(&k);
+                        // TODO: shouldn't blockhash also be deleted
```
Comment (CjS77) [inline base_layer/core/src/chain_storage/memory_db.rs]: I think you're right here.

File: base_layer/core/src/chain_storage/test/chain_backend.rs [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+    test_utils::builders::{create_test_block, create_test_kernel, create_utxo},
+    tx,
+    types::HashDigest,
+};
+use tari_mmr::MutableMmr;
+use tari_storage::lmdb_store::{db, LMDBBuilder};
+use tari_test_utils::paths::create_random_database_path;
+use tari_utilities::{hex::Hex, Hashable};
+
+pub fn create_lmdb_database() -> LMDBDatabase<HashDigest> {
```
Comment (CjS77) [inline base_layer/core/src/chain_storage/test/chain_backend.rs]: I feel like this function should actually be a utility function in `lmdb_db.rs` with a signature

`pub fn create_lmdb_backend(path: &Path) -> LMDBDatabase<HashDigest> {...}`

which allows the node software to decide where the database files should reside.

File: infrastructure/storage/src/lmdb_store/store.rs [hunk 3]
```diff
@@ -519,8 +527,8 @@ impl ReadOnlyIterator {
 }
 
 pub struct LMDBReadTransaction<'txn, 'db: 'txn> {
-    db: &'db Database<'db>,
-    access: ConstAccessor<'txn>,
+    pub db: &'db Database<'db>,
+    pub access: ConstAccessor<'txn>,
```
Comment (CjS77) [inline infrastructure/storage/src/lmdb_store/store.rs]: Unnecessary changes.

File: infrastructure/storage/src/lmdb_store/store.rs [hunk 4]
```diff
@@ -558,8 +566,8 @@ impl<'txn, 'db: 'txn> LMDBReadTransaction<'txn, 'db> {
 }
 
 pub struct LMDBWriteTransaction<'txn, 'db: 'txn> {
-    db: &'db Database<'db>,
-    access: WriteAccessor<'txn>,
+    pub db: &'db Database<'db>,
+    pub access: WriteAccessor<'txn>,
```
Comment (CjS77) [inline infrastructure/storage/src/lmdb_store/store.rs]: Unnecessary changes.

File: base_layer/core/src/chain_storage/error.rs [hunk 5]
```diff
@@ -66,4 +66,6 @@ pub enum ChainStorageError {
     MismatchedMmrRoot(MmrTree),
     // An invalid block was submitted to the database
     InvalidBlock,
+    // The Thread Safety has been breached and the data access has become poisoned
+    PoisonedAccess,
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/core/src/chain_storage/error.rs]: The `PoisonedAccess` error was removed and replaced with `AccessError(msg)`

File: base_layer/core/src/chain_storage/lmdb_db/lmdb.rs [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    bincode::serialize_into(&mut buf, data).map_err(|e| ChainStorageError::AccessError(e.to_string()))?;
+    Ok(buf)
+}
+
+pub fn deserialize<T>(buf_bytes: &[u8]) -> Result<T, error::Error>
+where T: DeserializeOwned {
+    bincode::deserialize(buf_bytes).map_err(|e| error::Error::ValRejected(e.to_string()))
+}
+
+pub fn lmdb_insert<K, V>(txn: &WriteTransaction, db: &Database, key: &K, val: &V) -> Result<(), ChainStorageError>
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/core/src/chain_storage/lmdb_db/lmdb.rs]: I made a note and will look at it in the next PR.

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 7]
```diff
@@ -0,0 +1,613 @@
...
+use lmdb_zero::{Database, Environment, WriteTransaction};
+use serde::{Deserialize, Serialize};
+use std::sync::{Arc, RwLock};
+use tari_mmr::{Hash as MmrHash, MerkleChangeTracker, MerkleCheckPoint, MerkleProof, MutableMmr};
+use tari_storage::lmdb_store::LMDBStore;
+use tari_utilities::hash::Hashable;
+
+type DatabaseRef = Arc<Database<'static>>;
+
+// TODO:  MerkleNode should be in its own file as it is used by PersistentDatabase and MemoryDatabase
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]: That does make sense. The comment was removed and the dependency on MerkleNode will be removed in the next PR.

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 8]
```diff
@@ -0,0 +1,613 @@
...
+{
+    env: Arc<Environment>,
+    metadata_db: DatabaseRef,
+    headers_db: DatabaseRef,
+    block_hashes_db: DatabaseRef,
+    utxos_db: DatabaseRef,
+    stxos_db: DatabaseRef,
+    kernels_db: DatabaseRef,
+    orphans_db: DatabaseRef,
+    utxo_mmr: RwLock<MerkleChangeTracker<D, LMDBVec<MmrHash>, Vec<MerkleCheckPoint> /* LMDBVec<MerkleCheckPoint> */>>,
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]: I looked and it seems like the MerkleChangeTracker needs to be changed to be able to get rid of the locks.

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 9]
```diff
@@ -0,0 +1,613 @@
...
+                            let hash = v.hash();
+                            lmdb_insert(&txn, &self.block_hashes_db, &hash, &k)?;
+                            let index = self
+                                .header_mmr
+                                .write()
+                                .map_err(|_| ChainStorageError::PoisonedAccess)?
+                                .push(&hash)? -
+                                1 as usize;
+                            let v = MerkleNode { index, value: *v };
+                            lmdb_insert(&txn, &self.headers_db, &k, &v)?;
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]: Agreed, I will look at removing unnecessary serialisations and deserializations in the next PR.

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 10]
```diff
@@ -0,0 +1,613 @@
...
+                                .map_err(|_| ChainStorageError::PoisonedAccess)?
+                                .push(&k)?;
+                            let proof_hash = v.proof().hash();
+                            let index = self
+                                .range_proof_mmr
+                                .write()
+                                .map_err(|_| ChainStorageError::PoisonedAccess)?
+                                .push(&proof_hash)? -
+                                1;
+                            let v = MerkleNode { index, value: *v };
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]: That is a great idea, will look at it in the next PR.

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 11]
```diff
@@ -0,0 +1,613 @@
...
+                            lmdb_insert(&txn, &self.kernels_db, &k, &v)?;
+                        },
+                        DbKeyValuePair::OrphanBlock(k, v) => {
+                            lmdb_insert(&txn, &self.orphans_db, &k, &v)?;
+                        },
+                        DbKeyValuePair::CommitBlock => {
+                            self.utxo_mmr
+                                .write()
+                                .map_err(|_| ChainStorageError::PoisonedAccess)?
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]: Change applied.

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 12]
```diff
@@ -0,0 +1,613 @@
...
+                            lmdb_insert(&txn, &self.utxos_db, &k, &v)?;
+                        },
+                        DbKeyValuePair::TransactionKernel(k, v) => {
+                            let index = self
+                                .kernel_mmr
+                                .write()
+                                .map_err(|_| ChainStorageError::PoisonedAccess)?
+                                .push(&k)? -
+                                1;
+                            let v = MerkleNode { index, value: *v };
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]: Unnecessary MerkleNode usages will be removed in next PR.

File: base_layer/core/src/chain_storage/test/chain_backend.rs (hunk 2 above)
Comment (neonknight64) [reply base_layer/core/src/chain_storage/test/chain_backend.rs]: Change applied, but modified signature slightly so it also returns a result.

File: infrastructure/storage/src/lmdb_store/store.rs (hunk 3 above)
Comment (neonknight64) [reply infrastructure/storage/src/lmdb_store/store.rs]: Good catch, change reverted.

File: infrastructure/storage/src/lmdb_store/store.rs (hunk 4 above)
Comment (neonknight64) [reply infrastructure/storage/src/lmdb_store/store.rs]: Change reverted.

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 13]
```diff
@@ -0,0 +1,613 @@
...
+                                .map_err(|_| ChainStorageError::PoisonedAccess)?
+                                .commit()
+                                .map_err(|e| ChainStorageError::AccessError(e.to_string()))?;
+                        },
+                    },
+                    WriteOperation::Delete(delete) => match delete {
+                        DbKey::Metadata(_) => {}, // no-op
+                        DbKey::BlockHeader(k) => {
+                            lmdb_delete(&txn, &self.headers_db, &k)?;
+                            // TODO: shouldn't blockhash also be deleted
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]: Will correct in next PR.

File: base_layer/core/src/chain_storage/memory_db.rs (hunk 1 above)
Comment (neonknight64) [reply base_layer/core/src/chain_storage/memory_db.rs]: Will correct in next PR.
